virttest/qemu_vm.py

```diff
@@ -2317,54 +2317,6 @@ def sort_key(dev):
         if bios_path:
             devices.insert(StrDev('bios', cmdline="-bios %s" % bios_path))
 
-        if params.get('ovmf_path'):
```

Comment:
Hi @PaulYuuu, @luckyh, this change code a test I had which was using code like:
```python
    vm.destroy(gracefully=True)
    if post_boot == "uefi":
        vm.params["ovmf_path"] = "/usr/share/edk2/ovmf/"
        vm.params["ovmf_code_filename"] = "OVMF_CODE.fd"
        vm.params["ovmf_vars_filename"] = "OVMF_VARS.fd"
    elif post_boot == "bios":
        vm.params["ovmf_path"] = ""
        vm.params["ovmf_code_filename"] = ""
        vm.params["ovmf_vars_filename"] = ""
    else:
        raise exceptions.TestError(f"Incorrect boot mode {post_boot}, must be 'uefi' or 'bios'")
    vm.create()
```
which reboots a vm from BIOS mode into UEFI mode and vice versa. Do you have any suggestion what should be the new approach for implementing the same after this PR?

Reply:
Hello @pevogam , could you help provide the debug outputs, please? If I remember correctly, when ovmf_path is empty, the pflash related command lines will not be appended.

Reply:
Hi @PaulYuuu,

> If I remember correctly, when ovmf_path is empty, the pflash related command lines will not be appended.

In the code snipped above I am adding the `ovmf_path` and rebooting a vm from bios to uefi, however the two lines never get added now that the qemu vm creation command is never checking for the `ovmf_path`. Is there an alternative way to implement the same code snippet which is compatible with your changes?

> Hello @pevogam , could you help provide the debug outputs, please?

I am not sure which debug output you need, could you be more specific? As I said above the lack of this code simply doesn't add the two pflash devices to the generated qemu command line. Here are debug output from before
```
2021-02-09 10:38:40,248 cpu              L0854 DEBUG| Detected CPU vendor as 'AuthenticAMD'
2021-02-09 10:38:40,353 qemu_vm          L2341 DEBUG| qemu will run in KVM mode
2021-02-09 10:38:40,544 qemu_vm          L2916 DEBUG| Devices of mis1: [t'PREFIX',t'qemu',t'-S',t'vmname',t'machine',t'i440FX',t'PIIX4_PM',t'PIIX3',t'piix3-ide',t'fdc',t'nodefaults',a'virtio-vga',t'mem',t'smp',t'cpu',hmp_id_hmp1,t'HMP-hmp1',hmp_id_catch_monitor,t'HMP-catch_monitor',chardev_serial0,serial0,t'isa-log',isdnhub,usb-tablet1,drive_image1,image1,idpRnmui,t'NET-bridge',id7Jb1WA,t'NET-bridge',idmGHsql,t'NET-bridge',idfzeZyB,t'NET-bridge',t'tcp-redir',t'display',t'rtc',t'bootmenu',t'kvm']
2021-02-09 10:38:40,544 qemu_vm          L2917 DEBUG| Buses of mis1
  drive_image1(QDrive): {drive:image1}
  isdnhub.0(uhci): {1:usb-tablet1}
  floppy(floppy): [None,None]
  ide(ide): [None,None,None,None]
  vcpu_bus(EPYC): {}
  _PCI_CHASSIS_NR(None): {}
  _PCI_CHASSIS(None): {}
  pci.0(PCI): {0x0-0x0:t'i440FX',0x1-0x0:t'PIIX3',0x1-0x1:t'piix3-ide',0x1-0x3:t'PIIX4_PM',0x2-0x0:a'virtio-vga',0x3-0x0:isdnhub,0x4-0x0:image1,0x5-0x0:idpRnmui,0x6-0x0:id7Jb1WA,0x7-0x0:idmGHsql,0x8-0x0:idfzeZyB}
2021-02-09 10:38:40,545 qemu_vm          L2995 INFO | Running qemu command (reformatted):
/usr/bin/qemu-kvm \
    -S  \
    -name 'mis1' \
    -machine pc  \
    -nodefaults \
    -device virtio-vga,bus=pci.0,addr=0x2 \
    -m 1024  \
    -smp 2,maxcpus=2,cores=1,threads=1,dies=1,sockets=2  \
    -cpu 'EPYC' \
    -chardev socket,id=hmp_id_hmp1,nowait,server,path=/tmp/monitor-hmp1-20210209-103838-dq3TSMYt  \
    -mon chardev=hmp_id_hmp1,mode=readline \
    -chardev socket,id=hmp_id_catch_monitor,nowait,server,path=/tmp/monitor-catch_monitor-20210209-103838-dq3TSMYt  \
    -mon chardev=hmp_id_catch_monitor,mode=readline \
    -chardev socket,id=chardev_serial0,nowait,server,path=/tmp/serial-serial0-20210209-103838-dq3TSMYt \
    -device isa-serial,id=serial0,chardev=chardev_serial0  \
    -chardev socket,id=seabioslog_id_20210209-103838-dq3TSMYt,path=/tmp/seabios-20210209-103838-dq3TSMYt,server,nowait \
    -device isa-debugcon,chardev=seabioslog_id_20210209-103838-dq3TSMYt,iobase=0x402 \
    -device piix3-usb-uhci,id=isdnhub,bus=pci.0,addr=0x3 \
    -device usb-tablet,id=usb-tablet1,bus=isdnhub.0,port=1 \
    -drive id=drive_image1,if=none,cache=none,format=qcow2,file=/mnt/local/images/c36_mis1/BusinessServer_64.qcow2 \
    -device virtio-blk-pci,id=image1,drive=drive_image1,bus=pci.0,addr=0x4 \
    -device virtio-net-pci,mac=02:00:00:00:01:01,id=idpRnmui,netdev=idnmnaeH,bus=pci.0,addr=0x5  \
    -netdev tap,id=idnmnaeH,fd=35 \
    -device virtio-net-pci,mac=02:00:00:00:01:02,id=id7Jb1WA,netdev=idgXyDQZ,bus=pci.0,addr=0x6  \
    -netdev tap,id=idgXyDQZ,fd=49 \
    -device virtio-net-pci,mac=02:00:00:00:01:03,id=idmGHsql,netdev=idtHwar7,bus=pci.0,addr=0x7  \
    -netdev tap,id=idtHwar7,fd=52 \
    -device virtio-net-pci,mac=02:00:00:00:01:04,id=idfzeZyB,netdev=idXQ1Ej4,bus=pci.0,addr=0x8  \
    -netdev tap,id=idXQ1Ej4,fd=57  \
    -vnc :0  \
    -rtc base=utc,clock=host  \
    -boot menu=off,order=cdn,once=dcn \
    -enable-kvm
    ```
and after the vm is recreated
```
2021-02-09 10:39:28,152 cpu              L0854 DEBUG| Detected CPU vendor as 'AuthenticAMD'
2021-02-09 10:39:28,257 qemu_vm          L2341 DEBUG| qemu will run in KVM mode
2021-02-09 10:39:28,465 qemu_vm          L2916 DEBUG| Devices of mis1: [t'PREFIX',t'qemu',t'-S',t'vmname',t'machine',t'i440FX',t'PIIX4_PM',t'PIIX3',t'piix3-ide',t'fdc',t'nodefaults',a'virtio-vga',t'mem',t'smp',t'cpu',hmp_id_hmp1,t'HMP-hmp1',hmp_id_catch_monitor,t'HMP-catch_monitor',chardev_serial0,serial0,t'isa-log',isdnhub,usb-tablet1,drive_image1,image1,idpRnmui,t'NET-bridge',id7Jb1WA,t'NET-bridge',idmGHsql,t'NET-bridge',idfzeZyB,t'NET-bridge',t'tcp-redir',t'display',t'rtc',t'bootmenu',t'kvm']
2021-02-09 10:39:28,466 qemu_vm          L2917 DEBUG| Buses of mis1
  drive_image1(QDrive): {drive:image1}
  isdnhub.0(uhci): {1:usb-tablet1}
  floppy(floppy): [None,None]
  ide(ide): [None,None,None,None]
  vcpu_bus(EPYC): {}
  _PCI_CHASSIS_NR(None): {}
  _PCI_CHASSIS(None): {}
  pci.0(PCI): {0x0-0x0:t'i440FX',0x1-0x0:t'PIIX3',0x1-0x1:t'piix3-ide',0x1-0x3:t'PIIX4_PM',0x2-0x0:a'virtio-vga',0x3-0x0:isdnhub,0x4-0x0:image1,0x5-0x0:idpRnmui,0x6-0x0:id7Jb1WA,0x7-0x0:idmGHsql,0x8-0x0:idfzeZyB}
2021-02-09 10:39:28,466 qemu_vm          L2995 INFO | Running qemu command (reformatted):
/usr/bin/qemu-kvm \
    -S  \
    -name 'mis1' \
    -machine pc  \
    -nodefaults \
    -device virtio-vga,bus=pci.0,addr=0x2 \
    -m 1024  \
    -smp 2,maxcpus=2,cores=1,threads=1,dies=1,sockets=2  \
    -cpu 'EPYC' \
    -chardev socket,id=hmp_id_hmp1,nowait,server,path=/tmp/monitor-hmp1-20210209-103838-dq3TSMYt  \
    -mon chardev=hmp_id_hmp1,mode=readline \
    -chardev socket,id=hmp_id_catch_monitor,nowait,server,path=/tmp/monitor-catch_monitor-20210209-103838-dq3TSMYt  \
    -mon chardev=hmp_id_catch_monitor,mode=readline \
    -chardev socket,id=chardev_serial0,nowait,server,path=/tmp/serial-serial0-20210209-103838-dq3TSMYt \
    -device isa-serial,id=serial0,chardev=chardev_serial0  \
    -chardev socket,id=seabioslog_id_20210209-103838-dq3TSMYt,path=/tmp/seabios-20210209-103838-dq3TSMYt,server,nowait \
    -device isa-debugcon,chardev=seabioslog_id_20210209-103838-dq3TSMYt,iobase=0x402 \
    -device piix3-usb-uhci,id=isdnhub,bus=pci.0,addr=0x3 \
    -device usb-tablet,id=usb-tablet1,bus=isdnhub.0,port=1 \
    -drive id=drive_image1,if=none,cache=none,format=qcow2,file=/mnt/local/images/c36_mis1/BusinessServer_64.qcow2 \
    -device virtio-blk-pci,id=image1,drive=drive_image1,bus=pci.0,addr=0x4 \
    -device virtio-net-pci,mac=02:00:00:00:01:01,id=idpRnmui,netdev=idnmnaeH,bus=pci.0,addr=0x5  \
    -netdev tap,id=idnmnaeH,fd=68 \
    -device virtio-net-pci,mac=02:00:00:00:01:02,id=id7Jb1WA,netdev=idgXyDQZ,bus=pci.0,addr=0x6  \
    -netdev tap,id=idgXyDQZ,fd=71 \
    -device virtio-net-pci,mac=02:00:00:00:01:03,id=idmGHsql,netdev=idtHwar7,bus=pci.0,addr=0x7  \
    -netdev tap,id=idtHwar7,fd=3 \
    -device virtio-net-pci,mac=02:00:00:00:01:04,id=idfzeZyB,netdev=idXQ1Ej4,bus=pci.0,addr=0x8  \
    -netdev tap,id=idXQ1Ej4,fd=5  \
    -vnc :0  \
    -rtc base=utc,clock=host  \
    -boot menu=off,order=cdn,once=dcn \
    -enable-kvm
```
The two pflash devices are never added despite the `ovmf_path` parameter being provided. Perhaps you could reproduce this with simpler code than the one above like:
```python
    vm.destroy(gracefully=True)
    vm.params["ovmf_path"] = "/usr/share/edk2/ovmf/"
    vm.params["ovmf_code_filename"] = "OVMF_CODE.fd"
    vm.params["ovmf_vars_filename"] = "OVMF_VARS.fd"
    vm.create()
```

Reply:
So I guess it might be also that I am misunderstanding something here. How do we register a qemu device when creating the vm by just providing it with a parameter like ` params["ovmf_path"]`? From the code changes I see in your PR you only remove any handling of such variable in the `make_create_command()`.

Reply:
Well, I think I know the root cause. When you using `ovmf_path`, the machine_type is still `pc`, but IIUC, qemu recommends the combination of `OVMF` + `q35`, the test case failed due to the `pc/i440FX` machine will not add pflash files, but q35 will.
https://github.com/avocado-framework/avocado-vt/blob/de0996188033f2f9143c4c7df1d9920d50514b6a/virttest/qemu_devices/qcontainer.py#L1190
https://github.com/avocado-framework/avocado-vt/blob/de0996188033f2f9143c4c7df1d9920d50514b6a/virttest/qemu_devices/qcontainer.py#L1124-L1141

@pevogam , could you use q35 machine type and try again?
In fact, I force on ppc64le and aarch64 platforms, so if I am wrong, please correct me, thanks.

Reply:
Sure, will try this next. Thanks for your advice @PaulYuuu!

Reply:
Hmmm, I changed the vm type to the more modern q35 at least to test for this particular code and the vm is installed, and initially handled correctly but fails on the `vm.create()` command above like:
```
2021-02-09 11:32:04,309 utils_misc       L0267 ERROR| Fail to create qemu command:
2021-02-09 11:32:04,312 utils_misc       L0272 ERROR| Original Traceback (most recent call last):
  File "/usr/lib/python3.8/site-packages/virttest/qemu_vm.py", line 2914, in create
    self.devices, self.spice_options = self.make_create_command()
  File "/usr/lib/python3.8/site-packages/virttest/qemu_vm.py", line 1503, in make_create_command
    devs = devices.machine_by_params(params)
  File "/usr/lib/python3.8/site-packages/virttest/qemu_devices/qcontainer.py", line 1395, in machine_by_params
    devices = machine_q35(cmd)
  File "/usr/lib/python3.8/site-packages/virttest/qemu_devices/qcontainer.py", line 1110, in machine_q35
    pflash_devices, cmd = pflash_handler("ovmf", cmd)
  File "/usr/lib/python3.8/site-packages/virttest/qemu_devices/qcontainer.py", line 1042, in pflash_handler
    img_info = json.loads(qemu_storage.QemuImg(img_params,
  File "/usr/lib64/python3.8/json/__init__.py", line 341, in loads
    raise TypeError(f'the JSON object must be str, bytes or bytearray, '
TypeError: the JSON object must be str, bytes or bytearray, not NoneType
```

Reply:
I have commented about a potential fix of this handling above, the question about enforcing q35 for UEFI mode remains though. Here is the screen that I see of when installing a UEFI enabled vm if I move from `pc` to `q35`:
![image](https://user-images.githubusercontent.com/1969979/107359430-55849100-6add-11eb-9c4d-be5fcb7e53d4.png)

Reply:
I only got `ovmf_path` in the variant of q35 before, and ovmf only works on x86 platform, so I considered binding it into the machine type definition, to avoid the wrong usage from users.
https://github.com/avocado-framework/avocado-vt/blob/de0996188033f2f9143c4c7df1d9920d50514b6a/virttest/shared/cfg/guest-hw.cfg#L464

However, as you mentioned above, `pc` can also launch a guest with UEFI, I am not sure it's the right way, but I can implement compatibility in `pc` if it's necessary.

Reply:
Before this PR I managed to use `x86_64` with OVMF so I guess you mean `x86` in general but yes I think that `pc` + `OVMF` is still a valid use (not wrong usage from users) despite the recommendation for q35. If you have any clue about the screen I got above perhaps I could move straight to q35 to resolve this at least for me (even though the general user might still encounter the errors I described). All I did to get the screen above was change the machine type to `q35` on a previously UEFI machine.